Adding team deleted event

efischer19 opened the pull request:
## [TNL-3199](https://openedx.atlassian.net/browse/TNL-3199)

This relies on @peter-fogg's delete changes currently living in the big "instructor tools" PR, https://github.com/edx/edx-platform/pull/9520, so this is currently branched from there for easier diff viewing. Once that change is merged into master, this will follow. I have no plans of including this change in that PR before it merges.

efischer19 commented:
I believe this is ready for reviewing, @dianakhuang and @peter-fogg could you take a look when you have a minute?

I based most of what I was doing off of https://github.com/edx/edx-platform/pull/9607

efischer19 commented:
Ah, I see from https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/3826/testReport/ that I'm incorrectly expecting the event to be emitted on invalid 401/403 requests. I'll get that fixed up.

peter-fogg commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -1039,10 +1043,22 @@ def test_delete_team(self, role):
+            {
+                'event_type': 'edx.team.deleted',
+                'event': {
+                    'course_id': self.course_id
```
This should check for `team_id` as well.

peter-fogg commented:
:+1: other than checking for `team_id` in the Bok Choy tests.

dianakhuang commented on lms/djangoapps/teams/tests/test_views.py:
```diff
@@ -509,7 +516,7 @@ def test_expand_public_user(self):
         ('Island', ['Search']),
-        ('search queryable', []),
+        ('not-a-query', []),
         ('team', ['Another Team', 'Public Profile Team']),
```
One thing I'd like to see here is verification that `leaner_removed` events also get fired here. You may need to add those events in separately.

efischer19 replied:
Oh yeah, good point. I hadn't even considered that. Since this PR is already based off the instructor tools PR, I'll go ahead and include the relevant types of `learner_removed` events:
 * Team is deleted, all memberships removed
 * Instructor uses tools to remove a membership

That doesn't touch "a learner leaves the team", but I could add that here too without too much trouble if you think that's the way to go.

efischer19 replied:
Also just to clarify - the line you commented on here isn't actually part of this PR, it just got caught up in the rebasing I was doing. Did you mean to add something to https://github.com/edx/edx-platform/pull/9613?

dianakhuang replied:
I meant to comment on this PR, but you might have been making changes, so Github got confused and moved some things around. I did cover 'a learner leaves the team' already, but the two situations you called out were not. Thanks for taking care of that!

efischer19 replied:
Awesome, let's keep on rolling. Thanks for the clarification.

efischer19 commented:
This should be ready for another round of review. Changes since last time are the inclusion on `learner_removed` events on delete team, and tests to cover that as well as the `learner_removed` events that are expected when using the edit membership page.

I made a small behavioral change to DELETE on MembershipDetailView to accommodate what @explorerleslie and I were discussing earlier. I'm not sure how "cleanly pythonic" this approach is, but it passes the tests so that's a start. https://github.com/edx/edx-platform/commit/a4cce885f12fe931f0a9f0c08d36825c638399fd

dianakhuang commented on lms/djangoapps/teams/views.py:
```diff
@@ -584,9 +584,21 @@ def delete(self, request, team_id):
+                'course_id': unicode(team.course_id),
+                'remove_method': 'team_deleted',
+                'user_id': member.user_id
+            })
```
Are we actually removing/deleting these memberships elsewhere? Or is that just part of what happens when the Django ORM deletes the team itself?

peter-fogg replied:
@dianakhuang Foreign keys automagically get followed by the ORM, so any memberships (and hypothetical other models that reference a team) will be deleted here.

peter-fogg commented on lms/djangoapps/teams/views.py:
```diff
@@ -584,9 +584,21 @@ def delete(self, request, team_id):
         """DELETE /api/team/v0/teams/{team_id}"""
         team = get_object_or_404(CourseTeam, team_id=team_id)
         self.check_object_permissions(request, team)
+        memberships = list(CourseTeamMembership.get_memberships(team_id=team_id))
```
Minor nit: the explicit call to `list` isn't really necessary; that happens implicitly when looping over the queryset on line 595.

efischer19 replied:
Yeah, but by that time the `team.delete()` call on line 589 has changed the data behind the queryset to be a less-helpful empty list.

dianakhuang replied:
It might make sense to use `copy` instead of `list` to preserve the data at this point. That way your intentions are more explicit.

peter-fogg replied:
Ah, I see.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/edit_team_members.js:
```diff
@@ -85,7 +85,7 @@
                             $.ajax({
                                 type: 'DELETE',
-                                url: self.teamMembershipDetailUrl + username
+                                url: self.teamMembershipDetailUrl.concat(username,'?admin=true')
```
Nit: this would be a little cleaner with passing `data: 'admin=true'`.

peter-fogg commented:
:+1: other than some nitpicky stuff (and the test failures).

efischer19 commented:
Comments have been addressed, and I think I fixed the failing tests (fixed locally, as far as I can tell)

efischer19 commented:
Ah yeah, it is. Thanks for the catch

efischer19 commented:
jenkins run js

efischer19 commented on lms/djangoapps/teams/static/teams/js/views/edit_team_members.js:
```diff
@@ -85,7 +85,7 @@
                             $.ajax({
                                 type: 'DELETE',
-                                url: self.teamMembershipDetailUrl + username
+                                url: self.teamMembershipDetailUrl.concat(username, '?admin=true')
```
@peter-fogg I've elected to keep `admin=true` as a querystring instead of using the request body as you mentioned yesterday. Is that something you feel particularly strongly about?

Discussion at https://openedx.atlassian.net/wiki/display/TNL/Team+API

peter-fogg replied:
Yeah, I'm entirely fine with the querystring approach. My earlier was just about passing the data to jQuery in a different way, but I didn't mean to change which part of the request holds the data.

efischer19 commented on lms/djangoapps/teams/views.py:
```diff
@@ -584,9 +584,23 @@ def delete(self, request, team_id):
         team = get_object_or_404(CourseTeam, team_id=team_id)
         self.check_object_permissions(request, team)
+        # Note: list() forces the queryset to be evualuated before delete()
+        memberships = list(CourseTeamMembership.get_memberships(team_id=team_id))
```
@dianakhuang You mentioned using `copy()` yesterday - is adding this comment sufficient to indicate what's going on, or would it be better to `import copy` and do it that way?

dianakhuang replied:
I think you should import `copy` if you need to. some objects have a `.copy()` method that you can use to get the copy.

dianakhuang replied:
Also, it might be that copying doesn't work, because it's only copying the iterator, not the actual iterator pointers. In that case, `list()` is fine, especially with that comment.

dianakhuang commented:
:+1: from me once the tests pass.

efischer19 commented:
That last commit was just a rebase to avoid any possible merge conflicts, all tests previously passed so I'm going ahead with the merge.
